Added first cut at attribute support.

mbeckerle opened the pull request:
Test enhanced to use attributes and global attributeGroup definition and reference as well as local attribute declarations.

Does not detect, nor handle,  the fact that an attribute is allowed to have the exact same name as an element in XML Schema.

Currently does not distinguish optional/required attributes (they're all addNullable), and does not implement XSD default/fixed.

Note that attribute definitions in XSD come last in the element definition lexically; however, they come out first in Drill field order. This is consistent with the way XML text looks where the element tag carries attributes first, in the opening tag, then child elements follow after the opening tag.

mbeckerle committed:
Added first cut at attribute support.

Test enhanced to use attributes and global attributeGroup definition and reference as well as local attribute declarations.

Does not detect, nor handle,  the fact that an attribute is allowed to have the exact same name as an element in XML Schema.

Currently does not distinguish optional/required attributes (they're all addNullable), and does not implement XSD default/fixed.

Note that attribute definitions in XSD come last in the element definition lexically; however, they come out first in Drill field order.
This is consistent with the way XML text looks where the element tag carries attributes first, in the opening tag, then child elements follow after the opening tag.

mbeckerle commented on contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/xsd/DrillXSDSchemaVisitor.java:
```diff
@@ -117,7 +117,9 @@ public void onExitElement(XmlSchemaElement xmlSchemaElement, XmlSchemaTypeInfo x
   @Override
   public void onVisitAttribute(XmlSchemaElement xmlSchemaElement, XmlSchemaAttrInfo xmlSchemaAttrInfo) {
-    // no op
+    String fieldName = xmlSchemaAttrInfo.getAttribute().getName();
```
Ah. this change breaks testSimpleXSD, so I'll have to fix that.

mbeckerle committed:
Fix testSimpleXSD

cgivre commented:
@mbeckerle Thanks for this.  One thing to note... The current implementation of the XML reader in Drill adds a map called `atttributes` for the attributes.  The key is the path for the attribute in question and they are separated by an underscore.

Take a look here:  https://github.com/apache/drill/tree/master/contrib/format-xml#attributes

That way you can have attributes with the same name as a key.  

Regarding the required/optional, mainly out of laziness, for the XSD reader, I made all the fields nullable in Drill.  I know that XSD supports required and optional fields, so we can certainly add that capability.  The tricky part is if you define that schema, and then you encounter data that doesn't follow that schema, the queries will fail.  My preference in Drill is to make it as fault tolerant as possible, so that the user gets something other than a stack trace...

mbeckerle commented:
> @mbeckerle Thanks for this. One thing to note... The current implementation of the XML reader in Drill adds a map called `atttributes` for the attributes. The key is the path for the attribute in question and they are separated by an underscore.
> 
> Take a look here: https://github.com/apache/drill/tree/master/contrib/format-xml#attributes
> 
> That way you can have attributes with the same name as a key.
> 
> Regarding the required/optional, mainly out of laziness, for the XSD reader, I made all the fields nullable in Drill. I know that XSD supports required and optional fields, so we can certainly add that capability. The tricky part is if you define that schema, and then you encounter data that doesn't follow that schema, the queries will fail. My preference in Drill is to make it as fault tolerant as possible, so that the user gets something other than a stack trace...

I will add the attributes to the attributes map.

As for the nullable aspect. I think at runtime of a query what you describe makes sense. I.e., if the document doesn't match the schema, then it should fall back on behavior that is like non-schema XML querying. However, when drill analyzes a query and there is an XSD, is it possible to issue a warning that a field is being requested that is not part of the schema? Many XML (and DFDL) data types are deeply nested. One will need help to formulate queries that are actually accessing the fields you want. If it just "always works but produces nothing", that will actually be quite hard to deal with.

cgivre commented:
@mbeckerle, 
Let me clarify a bit...

Drill is officially schema free.  However over the years we've discovered that reading data with no schema is a REALLY hard problem and lends itself to all kinds of issues that then get dumped on the user.  If you read through the Drill dev list you can see that we've had a lot of discussions around this topic.   This is in stark contrast with other federated query engines which require a schema in order to read data.

With that said, Drill settled on the approach of giving developers three options:
1.  Predefined schema:  This is useful for databases and other data sources where it is relatively easy to get schema metadata before the data is actually read.  In this case, you can define the schema and all the readers, and know that it will not change.  
2. "Late" schema:  This is a sort of middle ground.  In this case, you use whatever metadata is available before reading starts to initialize a schema, however, as new fields are encountered, they are added to the schema. 
3. Schema as you go:   This is where the schema is discovered as the data is being read. 

We also added the ability for users to provide the schema in various ways.  Users can provide a schema at query time, or via a file.  

What I was thinking was that once the XSD reader is working to our satisfaction, we'd enable the XML format plugin to download the XSD (cache it) and convert that to a schema.   I wouldn't use this to "lock" the schema, so that if Drill encounters data that is not defined in the schema, Drill could add that to the schema dynamically. 

Take a look at this PR:  https://github.com/apache/drill/pull/2710

Thank you for all your work on this!!

mbeckerle committed:
Changed attributes to use an enclosing element.

Added test of complex element with only attributes.

Added numerous assertion checks to be sure invariants are being maintained.

mbeckerle commented:
@cgivre  
So, given an XSD then, how would a Drill user's experience change? What does the XSD help them with?

I think given the nature of Drill, the right thing is for it to help them create correct queries, for when the XML documents in fact obey the schema. This means issuing warnings to them with informative instructions about what doesn't make sense, given the XSD schema. 

There are some asserts in the code where I wasn't sure what to do. E.g., what if the user's element declares a child element named "attributes" which conflicts with our reserved name?  What do we do in that situation?

mbeckerle commented on contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/xsd/DrillXSDSchemaVisitor.java:
```diff
@@ -59,10 +77,34 @@ public TupleMetadata getDrillSchema() {
+    //
+    // Note that the child name in constant ATTRIBUTE_MAP_NAME is reserved and cannot be used
+    // by any child element.
+    //
```
E.g., right here: what is the right way to error out here.

cgivre replied:
Something like this:

```java
throw UserException
        .dataReadError(e)
        .message(String.format("Failed to open input file: %s", file.split().getPath().toString()))
        .addContext(errorContext)
        .addContext(e.getMessage())
        .build(logger);
```

The `UserException` class has various errors which you can throw.   Also when used in the context of a file reader, you have another object called the `errorContext` which the file reader populates.  This actually contains a lot of useful information such as the file path, storage plugin etc.  It was all part of a greater effort to provide more helpful error messages.

Passing the `errorContext` is not strictly necessary, but helpful for the users to debug what went wrong.

mbeckerle replied:
What is the right ErrorType for a Schema error? I.e, the data source uses schemas (of some sort), and there is something wrong with the schema, so it cannot be used/work, uses an unsupported feature of the schema language, or the schema is simply malformed?

Should I create one?

cgivre replied:
In the `UserException` there is a `schemaChangeException` which is usually used to indicate that the schema changed rather than it is malformed.  You could certainly add a `malformedSchemaException` or something like that.   The `UserException` class does have some other errors.  If you decide to add a new error, I'd do it in the `UserException` class.

mbeckerle commented:
Another comment w.r.t. schema and what it is used for.

The non-schema-aware XML-reader treats all content as strings. 

But with schema, we have real simple types like integers, double, float, date, time, etc.

What happens when incoming data doesn't convert to the expected type? Is there any sort of warning? Can the schema be "broadened" to change the type to 'string'? If the query operates on the data (doing arithmetic say), then that will likely fail at runtime in that case. 

In some sense, we have the information necessary to specify in the error message that the data field was expected to be a number (from the XML schema of the data), but was not in conformance with this.

cgivre commented:
> 

I'm on the road today but will respond to this tonight.

cgivre commented:
> 

@mbeckerle 
You've encountered another challenge that exists in Drill reading data without a schema.  
Let me explain a bit about this and I'm going to use the JSON reader as an example.  First Drill requires data to be homogeneous.   Drill does have a Union vector type which allows heterogeneous data however this is a bit experimental and I wouldn't recommend using it.  Also, it really just shifts schema inconsistencies to the user.

For instance, let's say you have a column consisting of strings and floats.  What happens if you try to do something like this:

```sql
SELECT sum(mixed_col)
-- or
SELECT.... ORDER BY mixed_col
```

Remembering that Drill is distributed and if you have a column with the same name and you try to do these operations, they will fail.  

Let's say we have data like this:

```json
[
  {
     'col1': 'Hi there',
     'col2': 5.0
  },
  {
     'col1':True,
     'col2': 4,
     'col3': 'foo'
  }
]
```
 
In older versions of Drill, this kind of data, this would throw all kinds of `SchemaChangeExceptions`.   However, in recent versions of Drill, @jnturton submitted https://github.com/apache/drill/pull/2638 which overhauled implicit casting.  What this meant for users is that `col2` in the above would be automatically cast to a FLOAT and `col1` would be automatically cast to a `VARCHAR`.  

However, when reading data the story is a little different.  What we did for the JSON reader was have several read modes.  The least tolerant attempts to infer all data types.  This seems like a great idea in practice, however when you start actually using Drill with real data, you start seeing the issues with this approach.  The JSON reader has a few configuration options that increase its tolerance for bad data.  The next level is `readAllNumbersAsDouble` which... as the name implies, reads all numeric data as Doubles and does not attempt to infer ints vs floats.  The next options is `allTextMode` which reads all fields as `VARCHAR`.  This should be used when the data is so inconsistent that it cannot be read with either mode.  These modes can be set globally, at the plugin level or at query time.

For the XML reader,  I didn't add type inference because I figured the data would be quite messy, however it wouldn't be that hard to add basically the same levels as the JSON reader. 

This fundamental issue exists in all the readers that read data without a schema.  My rationale for working on the XSD reader is that this will enable us to accurately read XML data with all the correct data types.

cgivre reviewed (approved):
LGTM +1
